1D Membrane Model for CO2 Capture and Utilization

Morgan88888888 opened the pull request:
## Summary/Motivation:
There is a new project ongoing to support NETL's reactive capture technology. The goal of this work is to integrate CO2-selective polymer membranes with electrochemical conversion to efficiently transform captured CO2 into formic acid. The unit and flowsheet models developed in the field of capturing and utilizing CO2 during this project will be maintained in this repository. These models will benefit other ongoing work, such as membrane design and process configuration optimization.

## Changes proposed in this PR:
- Created the workspace for ongoing CO2 Capture and Utilization modeling efforts to support NETL's reactive capture technology.
- Developed a one-dimensional membrane model for CO2 capture.
- Developed the model testing for 1D membrane model

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

andrewlee94 commented on idaes/models_extra/co2_capture_and_utilization/unit_models/membrane_1d.py:
```diff
@@ -0,0 +1,341 @@
+#################################################################################
+# The Institute for the Design of Advanced Energy Systems Integrated Platform
```
Pylint is reporting a number of issues with this file that should be addressed (it seems GitHub has added a new integration that shows these during code reviews).

andrewlee94 commented on idaes/models_extra/co2_capture_and_utilization/unit_models/membrane_1d.py:
```diff
@@ -0,0 +1,341 @@
+            doc="Bool indicating whether to include pressure balance for stream. Default=True.",
+        ),
+    )
+
```
Unnecessary line break.

andrewlee94 commented on idaes/models_extra/co2_capture_and_utilization/unit_models/tests/test_membrane_1d.py:
```diff
@@ -0,0 +1,254 @@
+from idaes.models.properties.modular_properties.base.generic_property import (
+    GenericParameterBlock,
+)
+from idaes.models_extra.ccus.unit_models import Membrane1D, MembraneFlowPattern
```
This name is not right `ccus` needs to be `co2_capture_and_utilization`.

andrewlee94 commented on idaes/models_extra/co2_capture_and_utilization/unit_models/membrane_1d.py:
```diff
@@ -0,0 +1,341 @@
+            "state, or if all flow is provided via mass transfer. Default=True.",
+        ),
+    )
+    CONFIG.declare(
```
You should have a line break here to separate `StreamCONFIG` from `CONFIG`.

andrewlee94 commented on idaes/models_extra/co2_capture_and_utilization/unit_models/membrane_1d.py:
```diff
@@ -0,0 +1,341 @@
+        ),
+    )
+    CONFIG.declare(
+        "sweep_flow",
```
This appears to duplicate the `has_feed` options in the `StreamCONFIG`. Do you actually need both (I suspect you only need this one)?

Morgan88888888 replied:
Thanks for pointing that out. You're right. However, I think "has_feed" is too generic. I'll keep "sweep_flow" there to maintain straightforward options.

andrewlee94 commented on idaes/models_extra/co2_capture_and_utilization/unit_models/membrane_1d.py:
```diff
@@ -0,0 +1,341 @@
+            "state, or if all flow is provided via mass transfer. Default=True.",
+        ),
+    )
+
```
Unneeded line breaks.

andrewlee94 commented on idaes/models_extra/co2_capture_and_utilization/unit_models/membrane_1d.py:
```diff
@@ -0,0 +1,341 @@
+    )
+
+    CONFIG.declare(
+        "flow_type",
```
This also appears to duplicate options from `StreamCONFIG` (again I suspect you only need this one).

Morgan88888888 replied:
I think it's more clear to have "finite_elements" in unit config level.

andrewlee94 commented on idaes/models_extra/co2_capture_and_utilization/unit_models/membrane_1d.py:
```diff
@@ -0,0 +1,341 @@
+        ),
+    )
+
+    CONFIG.declare(
```
Duplicated options again - in this case I think you want to keep the ones in `StreamCONFIG` to allow different component lists on each side.

Morgan88888888 replied:
Same, would like to keep an option in unit level.

andrewlee94 commented on idaes/models_extra/co2_capture_and_utilization/unit_models/membrane_1d.py:
```diff
@@ -0,0 +1,341 @@
+            self.config.sweep_side.has_feed = False
+
+        # Set flow directions
+        if self.config.flow_type == MembraneFlowPattern.cocurrent:
```
This and the above indicate you do not need these entries in `StreamCONFIG`. As a general rule, you should never programmatically change the input in a `Config` dict, as that is intended solely for user input. By changing the input, you are overriding input the user explicitly gave you, which will result in confusion.

Instead, you should remove these options from `StreamCONFIG` and set these inputs to the MSContactor directly below.

andrewlee94 commented on idaes/models_extra/co2_capture_and_utilization/unit_models/membrane_1d.py:
```diff
@@ -0,0 +1,341 @@
+                self.config.sweep_side.property_package = self.config.property_package
+
+        streams_dict = {
+            "feed_side": self.config.feed_side,
```
This is where you should set inputs directly.

E.g.
```
feed_dict = dict(**self.config.feed_side)
sweep_dict = dict(**self.config.sweep_side)

feed_dict["flow_direction"] = FlowDirection.forward  # Feed side will always have forward flow
if self.config.flow_type == MembraneFlowPattern.cocurrent:
    sweep_dict["flow_direction"] = FlowDirection.forward
else:
    sweep_dict["flow_direction"] = FlowDirection.backward
```

andrewlee94 commented on idaes/models_extra/co2_capture_and_utilization/unit_models/membrane_1d.py:
```diff
@@ -0,0 +1,341 @@
+        ),
+    )
+
+    for side_name in ["feed", "sweep"]:
```
Something to think about: "retentate" and "permeate" are more standard (and general) names for the sides of a membrane (especially for cases with no sweep stream).

Morgan88888888 replied:
Thanks for the insight! May use them in the future if needed.

andrewlee94 commented on idaes/models_extra/co2_capture_and_utilization/unit_models/membrane_1d.py:
```diff
@@ -0,0 +1,341 @@
+            "sweep_side": self.config.sweep_side,
+        }
+        self.mscontactor = MSContactor(
+            streams=streams_dict,
```
You can then do:
```
streams = {"feed_side": feed_dict, "sweep_side": sweep_dict}
```

andrewlee94 reviewed (changes requested):
Here is a first batch of comments. Most are minor things and ways to do things better, and otherwise what I have seen so far looks good.

Some general comments are:

1. Fix all the pylint issues (either by running `pylint` yourself or looking throguh the notifications here on GitHub).
2. Fix the tests so that they run.
3. We will also need some basic documentation to merge this PR.

andrewlee94 commented:
Also, Sphinx is complaining about a doc string in you model.  I think you need to add a line break at the end of line 57 in you model.

ksbeattie commented:
@Morgan88888888 any update on this?

Morgan88888888 commented:
> @Morgan88888888 any update on this?

 I am working on it and should be ready this week.

Morgan88888888 commented on idaes/models_extra/co2_capture_and_utilization/unit_models/membrane_1d.py:
```diff
@@ -0,0 +1,316 @@
+
+    def _make_geometry(self):
+
+        self.area = Var(initialize=100, units=units.cm**2, doc="The membrane area")
```
I agree, it's not ideal to hard code units. However, in this application, the common permeability data is in Barrer, and using centimeters aligns with it, especially in lab-scale settings.

andrewlee94 replied:
That is what the unit conversion tools are for, and we are generally interested in modelling at the pilot or commercial where m^2 is more applicable.

Morgan88888888 commented on idaes/models_extra/co2_capture_and_utilization/unit_models/membrane_1d.py:
```diff
@@ -0,0 +1,316 @@
+
+        self.permeance = Var(
+            self.flowsheet().time,
+            self.mscontactor.elements,
```
In the simple assumption, it isn't necessary. However, retaining it isn't particularly harmful either. 
The example you gave about indexing could occur in some membrane separations but isn't common in this application.  I'll keep this idea in mind for future improvements if the need arises.

andrewlee94 replied:
What happens here if the two property packages have different components? Should this instead use the intersection of the two component lists (which is easy to do)?

ksbeattie commented:
Pushing this to August.

ksbeattie commented:
@Morgan88888888 any news?  We're now on the Aug release cycle.

ksbeattie commented on idaes/models_extra/co2_capture_and_utilization/unit_models/__init__.py:
```diff
@@ -0,0 +1,13 @@
+# Framework (IDAES IP) was produced under the DOE Institute for the
+# Design of Advanced Energy Systems (IDAES).
+#
+# Copyright (c) 2018-2023 by the software owners: The Regents of the
```
Copyright year range should be: 2018-2024

andrewlee94 replied:
Copyright still needs to be updated.

ksbeattie commented on idaes/models_extra/co2_capture_and_utilization/unit_models/membrane_1d.py:
```diff
@@ -0,0 +1,316 @@
+# Framework (IDAES IP) was produced under the DOE Institute for the
+# Design of Advanced Energy Systems (IDAES).
+#
+# Copyright (c) 2018-2023 by the software owners: The Regents of the
```
Copyright year range should be: 2018-2024

ksbeattie commented on idaes/models_extra/co2_capture_and_utilization/unit_models/tests/test_membrane_1d.py:
```diff
@@ -0,0 +1,257 @@
+# Framework (IDAES IP) was produced under the DOE Institute for the
+# Design of Advanced Energy Systems (IDAES).
+#
+# Copyright (c) 2018-2023 by the software owners: The Regents of the
```
Copyright year range should be: 2018-2024

ksbeattie commented:
@Morgan88888888 what's the news on this?

ksbeattie commented:
@Morgan88888888, we're closing this as it appears stalled.  Feel free to re-open if/when you are able to work on it again.

Morgan88888888 commented:
Based on the feedback, several major improvements have been made:

Documentation has been added to provide better model explanations.
Improved support for multi-property package applications.
Enhanced testing to cover various configurations.
Added functions to report stream tables.

Morgan88888888 commented:
> Also, Sphinx is complaining about a doc string in you model. I think you need to add a line break at the end of line 57 in you model.

A line break has been added, and more documentation has been included.

andrewlee94 commented on idaes/models_extra/co2_capture_and_utilization/unit_models/membrane_1d.py:
```diff
@@ -0,0 +1,316 @@
+One-dimensional membrane class for CO2 gas separation
+"""
+
+# pylint: disable=unused-import
```
Again, why is this disabled? This implies you have an unused import which should be removed.

Morgan88888888 replied:
It is disabled to pass the pylint test.

andrewlee94 replied:
You miss my point. I am asking why you are disabling the test, as you generally should not turn off pytest checks.

andrewlee94 commented on idaes/models_extra/co2_capture_and_utilization/unit_models/membrane_1d.py:
```diff
@@ -0,0 +1,321 @@
+            self.mscontactor.elements,
+            self.mscontactor.feed_side.component_list,
+            initialize=1,
+            doc="Values in Gas Permeance Unit(GPU)",
```
Missing space "Unit (GPU)"

andrewlee94 commented on idaes/models_extra/co2_capture_and_utilization/unit_models/membrane_1d.py:
```diff
@@ -0,0 +1,321 @@
+            doc=" This is a coefficient that will convert the unit of permeability from GPU to SI units for further calculation",
+        )
+
+        """
```
Pylint complains about this - general practice is to do this as single line comments (i.e. use `#`).

andrewlee94 commented on idaes/models_extra/co2_capture_and_utilization/unit_models/membrane_1d.py:
```diff
@@ -0,0 +1,316 @@
+        self.selectivity = Var(
+            self.flowsheet().time,
+            self.mscontactor.elements,
+            self.mscontactor.feed_side.component_list,
```
How are you handling the possible over-specification here?

andrewlee94 commented on idaes/models_extra/co2_capture_and_utilization/unit_models/membrane_1d.py:
```diff
@@ -0,0 +1,321 @@
+            return (
+                self.permeance[t, e, a] * self.selectivity[t, e, a, b]
+                == self.permeance[t, e, b]
+            )
```
This looks like it results in an over-specified problem. You write a constraint for `a, b` and `b, a` which are reflections of each other. If the user specifies all the selectivities they will have a degree of freedom issue.

I think you really need to only cover the unique pairs (and exclude a=b).

andrewlee94 commented on idaes/models_extra/co2_capture_and_utilization/unit_models/membrane_1d.py:
```diff
@@ -0,0 +1,321 @@
+
+        p_units = feed_side_units.PRESSURE
+
+        crossover_component_list = list(
```
I think you needed this earlier. It might also be a good idea to attach this ot the model to save Pyomo replicating this every time you use it.

Morgan88888888 replied:
moved and used for permeance as well

andrewlee94 commented on idaes/models_extra/co2_capture_and_utilization/unit_models/membrane_1d.py:
```diff
@@ -0,0 +1,321 @@
+            self.mscontactor.elements,
+            doc="Energy balance",
+        )
+        def energy_transfer(self, t, s):
```
For clarity, I would call this an isothermal constraint. It is not an energy balance.

andrewlee94 replied:
I would also suggest changing the name of the constraint (i.e. the name of the rule/function).

Morgan88888888 replied:
changed

andrewlee94 commented on idaes/models_extra/co2_capture_and_utilization/unit_models/membrane_1d.py:
```diff
@@ -0,0 +1,321 @@
+            {
+                "Feed Inlet": self.feed_side_inlet,
+                "Feed Outlet": self.feed_side_outlet,
+                "Permeate Inlet": self.sweep_side_inlet,
```
You need logic here to handle cases where the sweep stream is not present.

Morgan88888888 replied:
added

andrewlee94 reviewed (commented):
A number of comments on some edge cases and a few outstanding issues.

andrewlee94 commented:
@Morgan88888888 Please of not mark issues as resolved unless they are minor typo corrections. Reviewers need to be able to look over the issues and ensure that they have been addressed.

andrewlee94 commented:
@Morgan88888888 Also, this PR still needs to have at least basic documentation before it can be merged.

Morgan88888888 commented:
@andrewlee94 I have addressed your comments. I changed the model input to require permeance for all components across the membrane, and basic documentation has been added.

andrewlee94 commented on idaes/models_extra/co2_capture_and_utilization/unit_models/tests/test_membrane_1d.py:
```diff
@@ -66,17 +68,44 @@ def test_config():
+    )
+
+    # Check unit config arguments
+    print("=====================================================")
```
You should remove these print statements for the tests.

Morgan88888888 replied:
removed

andrewlee94 commented on docs/reference_guides/model_libraries/models_extra/membrane_model/1d_membrane.rst:
```diff
@@ -0,0 +1,14 @@
+can be connected for this application. The two sides of the membrane are called the feed side
+and sweep side. The sweep stream inlet is optional. The driving force across the membrane is the
+partial pressure difference in this gas separation application. Additionally, the energy balance
+assumes that temperature remains constant on each side of the membrane.
```
A few comments here:

1. Reference guides should be entirely autogenerated using Sphinx. Any thing that is hand written really should be an explanation (they go else where in the docs structure).
2. It would be good to have a short code example here, or a discussion of what the inputs/degrees of freedom are. This will help users understand what information they should be supplying.

Morgan88888888 replied:
Thanks and more info about model inputs/degrees of freedom are added.

andrewlee94 reviewed (commented):
A few more minor things.

andrewlee94 commented on docs/reference_guides/model_libraries/models_extra/membrane_model/1d_membrane.rst:
```diff
@@ -10,5 +10,30 @@ and sweep side. The sweep stream inlet is optional. The driving force across the
+* Feed flowrate - :math:`F_fr`
+* Feed compositions - :math:`x`
+* Feed pressure - :math:`P`
+* Feed temperature - :math:`T`
```
Should selectivity be here, or is that calculated from permeances?

Also, should there be mention of sweep conditions here too?

Morgan88888888 replied:
Thanks! The selectivity could be calculated and sweep condition is mentioned in the beginning.

andrewlee94 reviewed (commented):
One more minor comment for now. Otherwise I am just waiting to see that the tests all pass (so I suggest you not push any changes until they finish).

andrewlee94 reviewed (approved):
Pending fixing the final test failure, I think this is OK.
